python-lib/dku_stress_test_center/stress_test_center.py

```diff
@@ -1,166 +1,177 @@
+    def create(test_name, test_config):
+        test_class, test_type = DkuStressTestCenterConstants.TESTS[test_name]
+        params = test_config["params"]
+        if test_type == DkuStressTestCenterConstants.FEATURE_PERTURBATION:
+            feature_perturbation = test_class(**params)
+            features = test_config["selected_features"]
+            return FeaturePerturbationTest(feature_perturbation, features)
+        if test_type == DkuStressTestCenterConstants.SUBPOPULATION_SHIFT:
+            subpopulation_shift = test_class(**params)
+            return SubpopulationShiftTest(subpopulation_shift)
+        raise ValueError("Unknown stress test %s" % test_name)
 
-            else:
-                xt = pertubed_df.drop(target_column, axis=1).values
-                yt = pertubed_df[target_column].values
-                (xt, yt) = config.shift.transform(xt, yt)
+    def perturb_df(self, df: pd.DataFrame, target: str):
+        raise NotImplementedError()
 
-                pertubed_df.loc[:, pertubed_df.columns != target_column] = xt
+    def compute_metrics(self, perf_metric: str, clean_y_true: np.array, clean_y_pred: np.array,
+                        perturbed_y_true: np.array, perturbed_y_pred: np.array):
+        raise NotImplementedError()
 
-            pertubed_df.loc[:, target_column] = yt
 
-            pertubed_df[DkuStressTestCenterConstants.STRESS_TEST_TYPE] = stress_test_id
-            pertubed_df[DkuStressTestCenterConstants.DKU_ROW_ID] = pertubed_df.index
+class FeaturePerturbationTest(StressTest):
+    TEST_TYPE = DkuStressTestCenterConstants.FEATURE_PERTURBATION
 
-            # probably need to store the shifted_indices
+    def __init__(self, shift: Shift, features: list):
+        super(FeaturePerturbationTest, self).__init__(shift)
+        self.features = features
 
-            perturbed_datasets_df = perturbed_datasets_df.append(pertubed_df)
+    def perturb_df(self, df: pd.DataFrame, target: str):
+        X = df.loc[:, self.features].values
+        y = df.loc[:, target].values
 
-        return perturbed_datasets_df
+        X, y = self.shift.transform(X, y)
+        df.loc[:, self.features] = X
+        df.loc[:, target] = y
 
+        return df
 
-def build_stress_metric(y_true: np.array,
-                        y_pred: np.array,
-                        stress_test_indicator: np.array,
-                        pos_label: str or int):
-    # batch evaluation
-    # return average accuracy drop
-    # return robustness metrics: 1-ASR or imbalanced accuracy for prior shift
+    def compute_metrics(self, perf_metric: str, clean_y_true: np.array, clean_y_pred: np.array,
+                        perturbed_y_true: np.array, perturbed_y_pred: np.array):
+        return {
+            "robustness": stress_resilience(clean_y_pred, perturbed_y_pred),
+            "performance_variation": performance_variation(
+                perf_metric, clean_y_true, clean_y_pred, perturbed_y_true, perturbed_y_pred
```

Comment:
In FeaturePerturbationTest `clean_y_true` and `perturbed_y_true` should always been the same. Don't know if there's anything to do about that, though.

Reply:
ooh you're right, I was thinking `predicted_y_true` might be shorter (in the case where missing values are dropped) but actually this should be prevented by `clean_df_with_pred = self._clean_df.loc[test.df_with_pred.index, :]`